### PR TITLE
Clusterrmdup rewrite & improvement

### PR DESCRIPTION
**Pull request by FrickTobias**
Rewrote it to be more readable, issue #42.

Also noticed that some barcode duplicates were missed in previous version due to an assumption that read pairs come in order, which is not true in respect of the upstream mate to the current read. So I changed the script to parse read pair chunks whenever the current read position changes rather than whenever the read or mate positions changed.

### CODE REVIEW

**Comment by FrickTobias** [comment]
I'll look into the problem a bit more and reopen the PR when I've solved this.

**Comment by FrickTobias** [comment]
So it appear reads are only sorted on starting position in `samtools sort` and the tiebreaking somehow differs between linux and osx which affected the order in which reads were read and subsequently the read pair cache system in which they were parsed every time a new (read_start,read_stop) was encountered (for the downstream read of a pair).

I removed the read_stop and now the cache system parses chunks of reads every time a read_start is encountered for read pairs.

File: src/blr/cli/clusterrmdup.py [hunk 1]
```diff
@@ -3,594 +3,303 @@
...
 
-        import time
-        with open(self.log, 'a') as openout:
-            openout.write(time.strftime("%a, %d %b %Y %H:%M:%S", time.localtime()))
-            for objectVariable, value in vars(self).items():
-                openout.write('\n\n'+str(objectVariable) + '\n' + str(value))
 
 def add_arguments(parser):
     parser.add_argument("input_tagged_bam", help=".bam file tagged with RG tags and duplicates marked (not taking "
                                                  "cluster id into account).")
```
[the remark this answers is not in the document]
Comment (FrickTobias) [reply src/blr/cli/clusterrmdup.py]: Great catch, just writing it here for future documentation that the files does not need to be indexed anymore since the way of calculating summary stats was switched, so I will leave that part out.

File: src/blr/cli/clusterrmdup.py [hunk 2]
```diff
@@ -3,594 +3,303 @@
...
 logger = logging.getLogger(__name__)
 
 
 def main(args):
+    logger.info("Starting Analysis")
+
+    summary = Summary()
+
+    current_cache_rp = dict()
+    cache_read_pair_tracker = CacheReadPairTracker(None, None, None, None, None)
```
[the remark this answers is not in the document]
Comment (FrickTobias) [reply src/blr/cli/clusterrmdup.py]: That's true, it was needed in a previous version of the rewrite, I'll remove it as well as the `if read and mate` statement in the init function since this was only written to not run when read and mate was supplied as `None`.

File: src/blr/cli/clusterrmdup.py [hunk 3]
```diff
@@ -3,594 +3,303 @@
...
-            # Add all barcodes IDs to set, check later if total > 2 at positions.
-            if not readpair_pos_tuple in possible_duplicate_seeds:
-                possible_duplicate_seeds[readpair_pos_tuple] = set()
-            possible_duplicate_seeds[readpair_pos_tuple].add(barcode_ID)
+        if read and mate:
+            self.add_read_pair(read=read, mate=mate, bc=bc)
 
-        # When all overlaps found for current position, try finding proximal read pairs with same overlap
-        for possible_seed, barcode_IDs in possible_duplicate_seeds.items():  # Only one entry
+    def add_read_pair(self, read, mate, bc):
```
[the remark this answers is not in the document]
Comment (FrickTobias) [reply src/blr/cli/clusterrmdup.py]: On this point I am not sure if I agree. 

The read pair cache in its most basic component only tracks the read pairs for any given `read_pos_tuple` and it is only for convenience it has other functions and such build into it, such as a `.barcodes` variable (instead of calling on the barcodes in a for-loop over the read pairs). It was also a way to avoid using the `.get_tag()` function multiple times since bc is already fetched by prior to this step.

Although I see your point that it might be confusing and I guess the function does not need to track the reads itself now that it has a quite clear cut purpose as it would suffice to only track the relevant information about the read (in this case `rp_pos_tuple`, `barcodes`, `read_duplicate` and `mate_duplicate`).

In that case a more suitable name would maybe be something along the line of `barcode_set_cache`, which then has an ID according `rp_pos_tuple` and stores whether it has duplicates at both read/mate positions using those variables. What do you think of that?

File: src/blr/cli/clusterrmdup.py [hunk 4]
```diff
@@ -3,594 +3,303 @@
...
-            value_vector_dict[bc_id] = max(value_vector_dict[bc_id])
+    return bc_minimum
 
-        return value_vector_dict
 
-class BarcodeDuplicates:
+def reduce_several_step_redundancy(merge_dict):
     """
-    Tracks barcode ID:s which have readpairs ovelapping with others.
+    Takes translation  dict saved in object and makes sure 5->3, 3->1 becomes 5->1, 3->5
```
[the remark this answers is not in the document]
Comment (FrickTobias) [reply src/blr/cli/clusterrmdup.py]: Yes

**Comment by FrickTobias** [comment]
Thanks for the thorough feedback. I consented to everything, but one comment which I hope you'd give some further input on.

Regarding the definition I would only add that the 'at least two barcode' part has to be the same for the both positions. I usually call this that they 'share' at least two barcodes, but I believe this is what you meant. Anyway, I'd phrase it as:

- Two positions (positions defined as a unique set of `read_start`, `read_stop`, `mate_start`, `mate_stop`)) 
- at a maximum of W (`--window`, default 100kbp, between = `max(downstream_pos)-max(downstream_pos))` bp apart 
- sharing more than one barcode (share = `union(bc_set_pos1, bc_set_pos2)`).

The downside of using fragment boundaries is we might miss some merges. Since the rationale of merging barcodes is merging initial molecules since they actually are origin to the same molecule, but due to laboratorial imperfections does not appear as such. Having a generous window size (~3x molecule length) should be able to handle this with the downside of being prone for doing false merges. 

However, when investigating this by regulating the number of shared duplicates to call a barcode merge we found that increasing the required limit only seemed to decrease sensitivity while not affecting accuracy. So I think a way forward would to be to only require one overlapping read or read pair rather than pairs for seeding duplicates and a separate threshold for calling them (such as the overlapping reads in the molecules you suggest to incorporate into the analysis). But I suspect this might be a lot work without a lot of improvement to the overall picture of the dataset.

**Comment by pontushojer** [comment]
God to know that there has been tests on this! As you say this might not be relevant now but I figure its always good to think about improvements. 

I still have some comment that you have not replied to (as far as I can see). They are stated as "4 hidden conversations" in my browser so maybe that is why you missed them.

File: src/blr/cli/clusterrmdup.py [hunk 5]
```diff
@@ -3,594 +3,303 @@
...
+                    current_cache_rp = dict()
+                    current_cache_rp[rp_pos_tuple] = CacheReadPairTracker(rp_pos_tuple=rp_pos_tuple,
+                                                                   chromosome=chrom_new, read=read, mate=mate,
+                                                                   bc=bc_new)
+
+                # New chr => reset dup pos cache and rp cache tracker
+                else:
+                    cache_dup_pos = dict()
+                    for cache_read_pair_tracker in current_cache_rp.values():
+                        summary.reads_at_analyzed_dup_position += len(cache_read_pair_tracker.current_reads) * 2
```
[the remark this answers is not in the document]
Comment (FrickTobias) [reply src/blr/cli/clusterrmdup.py]: This statement was actually faulty, the analyzed reads should only be counted when they are sent to seed duplicates.

File: src/blr/cli/clusterrmdup.py [hunk 6]
```diff
@@ -3,594 +3,303 @@
...
+        rp_meet_requirements = False
 
-                # If this position has no reads yet, add position as key giving empty list as value
-                if not positions in singleton_duplicate_position:
-                    singleton_duplicate_position[chromosome][positions] = list()
+    try:
+        bc_new = int(read.get_tag(barcode_tag))
+    except KeyError:
+        summary.non_tagged_reads += 2
+        rp_meet_requirements = False
```
[the remark this answers is not in the document]
Comment (FrickTobias) [reply src/blr/cli/clusterrmdup.py]: That's true, possibly this points us to that we should add some non-tagged reads to test file so we'll get an error on this.

File: src/blr/cli/clusterrmdup.py [hunk 7]
```diff
@@ -3,594 +3,303 @@
...
+    cache_dup_pos = dict()
+    cache_reads = dict()
+    chrom_prev = None
+    merge_dict = dict()
+    with pysam.AlignmentFile(args.input_tagged_bam, "rb") as openin:
+        for read in tqdm(openin.fetch(until_eof=True), desc="Processing reads"):
+            summary.tot_reads += 1
+
+            # Wait for mate of read until process
+            if not read.query_name in cache_reads:
```
[the remark this answers is not in the document]
Comment (FrickTobias) [reply src/blr/cli/clusterrmdup.py]: That wasn't covered in Pycharms automatic reformatting command.

**Comment by FrickTobias** [comment]
Indeed I missed them, I read them and changed the code accordingly.

File: src/blr/cli/clusterrmdup.py [hunk 8]
```diff
@@ -77,8 +76,6 @@ def main(args):
                 # New chr => reset dup pos cache and rp cache tracker
                 else:
                     cache_dup_pos = dict()
-                    for cache_read_pair_tracker in current_cache_rp.values():
-                        summary.reads_at_analyzed_dup_position += len(cache_read_pair_tracker.current_reads) * 2
```
Comment (pontushojer) [inline src/blr/cli/clusterrmdup.py]: Why was this removed? Should you not analyze the duplicates for the last position on the previous chromosome?

Comment (FrickTobias) [reply src/blr/cli/clusterrmdup.py]: Turns out it could be corrected and simplified with the commit last commit.
